File: src/main/java/org/example/datnbbook/service/TacGiaService.java
package org.example.datnbbook.service;

import org.example.datnbbook.dto.TacGiaDTO;
import org.example.datnbbook.model.TacGia;
import org.example.datnbbook.repository.TacGiaRepository;
import org.springframework.beans.BeanUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class TacGiaService {
    @Autowired
    private TacGiaRepository tacGiaRepository;


    public List<TacGia> getAll() {
        return tacGiaRepository.findAllByDeletedFalseOrderByIdDesc();
    }

    public TacGia getById(Integer id) {
        return tacGiaRepository.findById(id)
                .filter(sp -> !sp.getDeleted())
                .orElseThrow(() -> new RuntimeException("Tác giả không tồn tại với id: " + id));
    }

    public TacGia create(TacGia sp) {
        if (sp.getMaTacGia() == null || sp.getMaTacGia().isEmpty()) {
            String nextMaTacGia = tacGiaRepository.getNextSequenceValue();
            sp.setMaTacGia(nextMaTacGia);
        }
        sp.setDeleted(false); // Đảm bảo Tác giả mới không bị đánh dấu xóa
        return tacGiaRepository.save(sp);
    }

    public TacGiaDTO update(Integer id, TacGiaDTO tacGiaDTO) {
        TacGia tacGia = tacGiaRepository.findById(id)
                .filter(sp -> !sp.getDeleted())
                .orElseThrow(() -> new RuntimeException("Tác giả không tồn tại với id: " + id));

        BeanUtils.copyProperties(tacGiaDTO, tacGia ,"maTacGia","id", "deleted");
        tacGia = tacGiaRepository.save(tacGia);
        return convertToDTO(tacGia);
    }

    public void delete(Integer id) {
        TacGia tacGia = tacGiaRepository.findById(id)
                .filter(sp -> !sp.getDeleted())
                .orElseThrow(() -> new RuntimeException("Tác giả không tồn tại với id: " + id));
        tacGia.setDeleted(true);
        tacGiaRepository.save(tacGia);
    }

    public List<TacGia> search(String keyword) {
        return tacGiaRepository.search(keyword);
    }

    private TacGiaDTO convertToDTO(TacGia tacGia) {
        TacGiaDTO tacGiaDTO = new TacGiaDTO();
        BeanUtils.copyProperties(tacGia, tacGiaDTO);
        return tacGiaDTO;
    }
}

File: src/frontend/vue-notus-main/vue-notus-main/vue-notus-main/src/service/TacGiaService.js
import api from './api';
import TacGia from '../models/TacGia';

class TacGiaService {
  async getAll() {
    const response = await api.get('/tac-gia');
    return response.data.map(item => new TacGia(item));
  }

  async getById(id) {
    const response = await api.get(`/tac-gia/${id}`);
    return new TacGia(response.data);
  }

  async create(tacGia) {
    const response = await api.post('/tac-gia', tacGia);
    return new TacGia(response.data);
  }

  async update(id, tacGia) {
    const response = await api.put(`/tac-gia/${id}`, tacGia.toJSON());
    return new TacGia(response.data);
  }

  async delete(id) {
    await api.delete(`/tac-gia/${id}`);
  }

  async search(keyword) {
    const response = await api.get('/tac-gia/search', {
      params: { keyword }
    });
    return response.data.map(item => new TacGia(item));
  }
}

export default new TacGiaService();